packages/cli/command_spec.pony

```diff
@@ -0,0 +1,278 @@
+    name = name'
+    descr = descr'
+    short = short'
+    (typ, default, required) = _init(I64Type, default')
+
+  new val f64(name': String, descr': String = "",
+    short': (U8 | None) = None, default': (F64 | None) = None) ?
+  =>
+    name = name'
+    descr = descr'
+    short = short'
+    (typ, default, required) = _init(F64Type, default')
+
+  // Other than bools, all options require args.
+  fun _requires_arg(): Bool =>
+    match typ |(let b: BoolType) => false else true end
+    // TODO: why can't Pony match on just type? |BoolType=>...
+
+  // Used for bool options to get the true arg when option is present w/o arg
+  fun _default_arg(): Value =>
+    match typ |(let b: BoolType) => true else false end
+
+  fun _has_short(sh: U8): Bool =>
+    match short
+    | let ss: U8 => sh == ss
+    else
+      false
+    end
+
+  fun deb_string(): String =>
+    "--" + name + "[" + typ.string() + "]" +
+      if not required then "(=" + default.string() + ")" else "" end
+
+  fun help_string(): String =>
+    let s = match short
+      | let ss: U8 => "-" + String.from_utf32(ss.u32()) + ", "
+      else
+        "    "
+      end
+    let l = "--" + name
+    s + l
+
+
+class val ArgSpec
+  """
+  ArgSpec describes the specification of a positional argument. Args always
+  come after a leaf command, and are assigned in their positional order.
+  """
+  let name: String
+  let descr: String
+  let typ: ValueType
+  let default: Value
+  let required: Bool
+
+  fun tag _init(typ': ValueType, default': (Value | None))
+  :
+    (ValueType, Value, Bool) ?
+  =>
+    match default'
+      | None =>
+        (typ', false, true)
+    else
+      (typ', default' as Value, false)
+    end
+
+  new val bool(name': String, descr': String="", default': (Bool|None)=None) ?
+  =>
+    name = name'
+    descr = descr'
+    (typ, default, required) = _init(BoolType, default')
+
+  new val string(name': String, descr': String="", default': (String|None)=None) ?
+  =>
+    name = name'
+    descr = descr'
+    (typ, default, required) = _init(StringType, default')
+
+  new val i64(name': String, descr': String="", default': (I64|None)=None) ?
+  =>
+    name = name'
+    descr = descr'
+    (typ, default, required) = _init(I64Type, default')
+
+  new val f64(name': String, descr': String="", default': (F64|None)=None) ?
+  =>
+    name = name'
+    descr = descr'
+    (typ, default, required) = _init(F64Type, default')
+
+  fun deb_string(): String =>
+    name + "[" + typ.string() + "]" +
+      if not required then "(=" + default.string() + ")" else "" end
+
+  fun help_string(): String =>
+    "<" + name + ">"
+
+
+type Value is (Bool | String | I64 | F64)
+
+primitive BoolType
+  fun string(): String => "Bool"
+
+primitive StringType
+  fun string(): String => "String"
+
+primitive I64Type
+  fun string(): String => "I64"
+
+primitive F64Type
+  fun string(): String => "F64"
+
+type ValueType is
+  ( BoolType
+  | StringType
+  | I64Type
+  | F64Type)
```

Comment:
Sure, I will double check. I think at one point I had the user passing these as type selectors but then I hid that behind the methods.